Let Model outputs be NumPy arrays

sdrave opened the pull request:
As discussed, outputs of Models are made NumPy arrays instead of VectorArrays to facilitate interfaces with pyMOR Models from user code / external code. 

**Edit:**
~I chose `output_dim` over `dim_output` to be consistent with the choice in `models.iosys` and also because I think it sounds nicer. However, this is somewhat inconsistent with the rest of pyMOR where 'dim_***' is used.~

The dimension of the output is given by `dim_output` attribute. `output_dim` of `IntputOutputModel` has been renamed accordingly. Its `input_dim` attribute has been renamed to `dim_input`. The old names are deprecated.

In addition, `input_space` and `output_space` have been removed from `InputOutputModel` and the `b` and `c` arrays of tangential directions for the interpolatory reductors have been made numpy arrays instead of `VectorArrays`.

sdrave commented:
@pmli, I'm not sure about the `rom0` stuff. Probably should now be NumPy arrays instead of `VectorArrays`?

HenKlei commented on src/pymor/models/iosys.py:
```diff
@@ -2034,7 +2030,7 @@ def __sub__(self, other):
         C = BlockRowOperator([self.C, -other.C])
         if self.D is other.D:
-            D = ZeroOperator(self.output_space, self.input_space)
+            D = ZeroOperator(self.D.range, self.D.range)
```
```suggestion
            D = ZeroOperator(self.D.range, self.D.source)
```

HenKlei commented on src/pymor/models/interface.py:
```diff
@@ -86,7 +89,10 @@ def output(self, mu=None, **kwargs):
         Returns
         -------
-        The computed model output as a |VectorArray| from `output_space`.
+        The computed model output as a as a 2D |NumPy array| with dimension
```
```suggestion
        The computed model output as a 2D |NumPy array| with dimension
```

pmli commented:
Removing `output_space` from `iosys` models feels like a big change and also like not enough of a big change. Using only `output_dim` seems like it implies that we always assume that the number of outputs is "small" and that the output space doesn't matter, which I'm not even sure how many implications it would have (e.g., it would change the assumptions methods like `eval_tf` need to make and using input/output space ids could be a way to implement a method for [interconnecting systems](https://www.mathworks.com/help/control/ref/connect.html)). On the other hand, if it's necessary to remove it, it would make sense to also remove `input_space`. I'm not sure of all the advantages and disadvantages. One good thing is that adding assumptions that `B` and `C` can always be converted to a `VectorArray` would make `eval_tf` more "pyMOR-ic".

sdrave commented:
> Removing `output_space` from `iosys` models feels like a big change and also like not enough of a big change. Using only `output_dim` seems like it implies that we always assume that the number of outputs is "small" and that the output space doesn't matter, which I'm not even sure how many implications it would have (e.g., it would change the assumptions methods like `eval_tf` need to make and using input/output space ids could be a way to implement a method for [interconnecting systems](https://www.mathworks.com/help/control/ref/connect.html)). 

I'm really confused. Isn't this what we discussed the entire time? When outputs are always NumPy arrays, then what's the point of defining an output space?

> On the other hand, if it's necessary to remove it, it would make sense to also remove `input_space`. 

Of course. I havn't done this, as this PR is only about the outputs and we havn't formalized inputs yet. It clear however, that when outputs are NumPy arrays, than inputs as well. (But we still need to discuss time, `Mu` and so on ..)

pmli commented:
> I'm really confused. Isn't this what we discussed the entire time?

I thought we only discussed editing the `solve` and `output` methods...

> When outputs are always NumPy arrays, then what's the point of defining an output space?

As I said, using space ids could be used to manage connecting inputs and outputs of different systems. In some situations, it would be natural to have two types of inputs/outputs, and then the input/output space could be a `BlockVectorSpace` where each subspace has a special id.

HenKlei commented on src/pymor/models/iosys.py:
```diff
@@ -221,7 +217,7 @@ def mag_plot(self, w, mu=None, ax=None, ord=None, Hz=False, dB=False, **mpl_kwar
-    def __init__(self, input_space, solution_space, output_space, cont_time=True,
+    def __init__(self, input_space, solution_space, output_dim, cont_time=True,
                  error_estimator=None, visualizer=None, name=None):
         super().__init__(input_space, output_space, cont_time=cont_time,
```
```suggestion
        super().__init__(input_space, output_dim, cont_time=cont_time,
```

sdrave commented:
> > I'm really confused. Isn't this what we discussed the entire time?
> 
> I thought we only discussed editing the `solve` and `output` methods...

`output_space` is part of `output`'s and `solve`'s api: it's the `VectorSpace` of the return values of these methods. If we change outputs to NumPy arrays but do not change `output_space` to `output_dim` you expose the user to a second notion of what the output seems to be. It is not clear to me, what the definition of this second notion actually would be and of what use it might be.

> 
> > When outputs are always NumPy arrays, then what's the point of defining an output space?
> 
> As I said, using space ids could be used to manage connecting inputs and outputs of different systems. In some situations, it would be natural to have two types of inputs/outputs, and then the input/output space could be a `BlockVectorSpace` where each subspace has a special id.

If I don't misunderstand this, the same thing could be easily achieved with an `output_ids` or `output_names` attribute. I am open to add something like this.

pmli commented:
> > > I'm really confused. Isn't this what we discussed the entire time?
> > 
> > 
> > I thought we only discussed editing the `solve` and `output` methods...
> 
> `output_space` is part of `output`'s and `solve`'s api: it's the `VectorSpace` of the return values of these methods. If we change outputs to NumPy arrays but do not change `output_space` to `output_dim` you expose the user to a second notion of what the output seems to be. It is not clear to me, what the definition of this second notion actually would be and of what use it might be.

Ok, so, for you, `C.range` and `D.range` are something internal to `LTIModel` and `output_space`/`output_dim` are something separate? I'm wondering if we remove `output_space` and only use `output_dim`, should then `C` be a `VectorArray`/`VectorArrayOperator`?

> > > When outputs are always NumPy arrays, then what's the point of defining an output space?
> > 
> > 
> > As I said, using space ids could be used to manage connecting inputs and outputs of different systems. In some situations, it would be natural to have two types of inputs/outputs, and then the input/output space could be a `BlockVectorSpace` where each subspace has a special id.
> 
> If I don't misunderstand this, the same thing could be easily achieved with an `output_ids` or `output_names` attribute. I am open to add something like this.

Ok, I guess that could work.

sdrave commented:
> > `output_space` is part of `output`'s and `solve`'s api: it's the `VectorSpace` of the return values of these methods. If we change outputs to NumPy arrays but do not change `output_space` to `output_dim` you expose the user to a second notion of what the output seems to be. It is not clear to me, what the definition of this second notion actually would be and of what use it might be.
> 
> Ok, so, for you, `C.range` and `D.range` are something internal to `LTIModel` and `output_space`/`output_dim` are something separate? 

Yes. I would go as far as saying that potentially having different `C.range` with the same `C.range.dim` is more of an annoyance than a feature if all outputs have to be NumPy arrays in the end. So one might even think about requiring that C.range always is a `NumpyVectorSpace`.

> I'm wondering if we remove `output_space` and only use `output_dim`, should then `C` be a `VectorArray`/`VectorArrayOperator`?

I think I originally would have liked `C` and `B` to be `VectorArrays`, but we decided against that at some point. Don't remember all the details. One consequence would be that they no longer could be parametric. Also, it is probably not what people would expect.

pmli commented:
> Yes. I would go as far as saying that potentially having different `C.range` with the same `C.range.dim` is more of an annoyance than a feature if all outputs have to be NumPy arrays in the end. So one might even think about requiring that C.range always is a `NumpyVectorSpace`.

Yes, I agree requiring that `C.range` (and `B.source`) are `NumpyVectorSpace` would be nice, but I'm not sure it would work for interconnected systems as in

![Control system](https://upload.wikimedia.org/wikipedia/commons/2/25/H-infty_plant_representation.png)

where system `P` has outputs `z` and `v`, and you would need to extract them somehow to assemble the interconnected system. I'm not sure that would be possible for a parametric `C` operator.

> I think I originally would have liked `C` and `B` to be `VectorArrays`, but we decided against that at some point. Don't remember all the details. One consequence would be that they no longer could be parametric. Also, it is probably not what people would expect.

Yes, it would definitely be good if `B` and `C` could be parametric.

sdrave commented:
> > Yes. I would go as far as saying that potentially having different `C.range` with the same `C.range.dim` is more of an annoyance than a feature if all outputs have to be NumPy arrays in the end. So one might even think about requiring that C.range always is a `NumpyVectorSpace`.
> 
> Yes, I agree requiring that `C.range` (and `B.source`) are `NumpyVectorSpace` would be nice, but I'm not sure it would work for interconnected systems as in
> 
> ![Control system](https://camo.githubusercontent.com/887a93c147a29671a132a9aa0a92ec1731bc6810/68747470733a2f2f75706c6f61642e77696b696d656469612e6f72672f77696b6970656469612f636f6d6d6f6e732f322f32352f482d696e6674795f706c616e745f726570726573656e746174696f6e2e706e67)
> 
> where system `P` has outputs `z` and `v`, and you would need to extract them somehow to assemble the interconnected system. I'm not sure that would be possible for a parametric `C` operator.

I am not quite sure how this is related to allowing `output_space` to be different from `NumpyVectorSpace`. If you are thinking of blocked spaces and assigning ids to the subspaces, as I said I am fine with assigning labels to every output component. That should give you enough information on how to connect things. If you want to interconnect `Models` to make a new `Model`, it should be trivial to write a new `Operator` that extracts a few components from a `NumpyVectorArray` and makes a new array from them.

> 
> > I think I originally would have liked `C` and `B` to be `VectorArrays`, but we decided against that at some point. Don't remember all the details. One consequence would be that they no longer could be parametric. Also, it is probably not what people would expect.
> 
> Yes, it would definitely be good if `B` and `C` could be parametric.

Ok, so `B` and `C` should stay `Operators`.

pmli commented:
> If you want to interconnect `Models` to make a new `Model`, it should be trivial to write a new `Operator` that extracts a few components from a `NumpyVectorArray` and makes a new array from them.

Yes, for a `NumpyMatrixOperator`, it should be easy to extract a few rows/columns, but I'm not sure if this is possible for a general parametric (not in parameter-separable form) `Operator`. Although, I guess this could be achieved by concatenating with a `ComponentProjectionOperator` (it might not be the most efficient thing, but it should work).

sdrave commented:
> > If you want to interconnect `Models` to make a new `Model`, it should be trivial to write a new `Operator` that extracts a few components from a `NumpyVectorArray` and makes a new array from them.
> 
> Yes, for a `NumpyMatrixOperator`, it should be easy to extract a few rows/columns, but I'm not sure if this is possible for a general parametric (not in parameter-separable form) `Operator`. Although, I guess this could be achieved by concatenating with a `ComponentProjectionOperator` (it might not be the most efficient thing, but it should work).

This is what I had in mind, actually. There might be a more convenient version where you can specify the names of the output components or so. Not sure about your concerns regarding efficiency. But there is also the option of having some version of `BlockOperator` which outputs stacked `NumpyVectorArrays` in case you want to dissect the operator itself.

sdrave commented:
I have now updated this PR to current master and added the following changes:

- As I agree that having `output_dim` but `input_space` is really strange, I also replaced `input_space` by `input_dim` for the sys-mor models.
- While I was at it, it seemed to me unnatural to have the tangential directions `b` and `c` in the interpolatory reductors as `VectorArrays`, so I changed them to numpy arrays.

Before merging this: Who prefers `dim_output` over `output_dim`, @pymor/all? Note that elsewhere we have `dim_source`, `dim_range`, `dim_domain`, so `dim_output` might actually make more sense.

ftalbrecht commented:
> Before merging this: Who prefers `dim_output` over `output_dim`, @pymor/all? Note that elsewhere we have `dim_source`, `dim_range`, `dim_domain`, so `dim_output` might actually make more sense.

Agreed

sdrave commented:
I just renamed `input_dim`/`output_dim` to `dim_input/dim_output`.

sdrave commented:
@pmli, this can now be merged, IMHO. Can you take a look?

pmli commented on src/pymor/reductors/h2.py:
```diff
@@ -66,16 +67,16 @@ def _check_rom0_params(self, rom0_params):
-            assert rom0_params['b'] in self.fom.input_space
-            assert rom0_params['c'] in self.fom.output_space
-            assert len(rom0_params['sigma']) == len(rom0_params['b'])
-            assert len(rom0_params['sigma']) == len(rom0_params['c'])
+            assert rom0_params['b'].shape[1] == self.fom.dim_input
+            assert rom0_params['c'].shape[1] == self.fom.dim_output
+            assert len(rom0_params['sigma']) == rom0_params['b'].shape[0]
+            assert len(rom0_params['sigma']) == rom0_params['c'].shape[0]
```
How about adding `assert isinstance(rom0_params['b'], np.ndarray)` above? (And similar for `'c'`.)

pmli commented on src/pymor/reductors/h2.py:
```diff
@@ -657,9 +654,9 @@ def _lti_to_poles_b_c(rom):
+        |NumPy array| of shape `(len(poles), rom.dim_input)`.
     c
-        |VectorArray| from `rom.C.range`.
+        |NumPy array| of shape `(len(poles), rom.dim_output)`.
```
I would prefer `rom.order` instead of `len(poles)`.

pmli commented on src/pymor/reductors/h2.py:
```diff
@@ -705,14 +702,14 @@ def _poles_b_c_to_lti(poles, b, c):
             A.append([[pole.real, pole.imag],
                       [-pole.imag, pole.real]])
-            bi = b[i].to_numpy()
+            bi = b[i]
             B.append(np.vstack([2 * bi.real, -2 * bi.imag]))
```
```suggestion
            B.append(np.vstack([2 * b[i].real, -2 * b[i].imag]))
```

pmli commented on src/pymor/reductors/h2.py:
```diff
@@ -705,14 +702,14 @@ def _poles_b_c_to_lti(poles, b, c):
+            bi = b[i]
             B.append(np.vstack([2 * bi.real, -2 * bi.imag]))
-            ci = c[i].to_numpy()
+            ci = c[i]
             C.append(np.hstack([ci.real.T, ci.imag.T]))
```
```suggestion
            C.append(np.hstack([c[i].real[:, np.newaxis], c[i].imag[:, np.newaxis]]))
```

pmli commented on src/pymor/reductors/h2.py:
```diff
@@ -705,14 +702,14 @@ def _poles_b_c_to_lti(poles, b, c):
-            B.append(b[i].to_numpy().real)
-            C.append(c[i].to_numpy().real.T)
+            B.append(b[i].real)
+            C.append(c[i].real.T)
```
I guess `c[i]` is now a 1D NumPy array.

```suggestion
            C.append(c[i].real[:, np.newaxis])
```

sdrave commented:
@renefritze, there are some strange new build failures here ..

renefritze commented:
> @renefritze, there are some strange new build failures here ..

Restart or rebase on #1158
